cloudinit/config/cc_chef.py

```diff
@@ -6,78 +6,286 @@
-
-**Supported distros:** all
-
-**Config keys**::
-
-    chef:
-       directories: (defaulting to /etc/chef, /var/log/chef, /var/lib/chef,
-                     /var/cache/chef, /var/backups/chef, /var/run/chef)
-       validation_cert: (optional string to be written to file validation_key)
-                        special value 'system' means set use existing file
-       validation_key: (optional the path for validation_cert. default
-                        /etc/chef/validation.pem)
-       firstboot_path: (path to write run_list and initial_attributes keys that
-                        should also be present in this configuration, defaults
-                        to /etc/chef/firstboot.json)
-       exec: boolean to run or not run chef (defaults to false, unless
-                                             a gem installed is requested
-                                             where this will then default
-                                             to true)
-
-    chef.rb template keys (if falsey, then will be skipped and not
-                           written to /etc/chef/client.rb)
-
-    chef:
-      client_key:
-      encrypted_data_bag_secret:
-      environment:
-      file_backup_path:
-      file_cache_path:
-      json_attribs:
-      log_level:
-      log_location:
-      node_name:
-      omnibus_url:
-      omnibus_url_retries:
-      omnibus_version:
-      pid_file:
-      server_url:
-      show_time:
-      ssl_verify_mode:
-      validation_cert:
-      validation_key:
-      validation_name:
-"""
+"""Chef: module that configures, starts and installs chef."""
 
 import itertools
 import json
 import os
+from textwrap import dedent
 
+from cloudinit.config.schema import (
+    get_schema_doc, validate_cloudconfig_schema)
 from cloudinit import templater
 from cloudinit import url_helper
 from cloudinit import util
+from cloudinit.settings import PER_ALWAYS
+
+
+frequency = PER_ALWAYS
+distros = ["all"]
+schema = {
+    'id': 'cc_chef',
+    'name': 'Chef',
+    'title': 'module that configures, starts and installs chef',
+    'description': dedent("""\
+        This module enables chef to be installed (from packages or
+        from gems, or from omnibus). Before this occurs chef configurations are
+        written to disk (validation.pem, client.pem, firstboot.json,
+        client.rb), and needed chef folders/directories are created
+        (/etc/chef and /var/log/chef and so-on). Then once installing proceeds
+        correctly if configured chef will be started (in daemon mode or in
+        non-daemon mode) and then once that has finished (if ran in non-daemon
+        mode this will be when chef finishes converging, if ran in daemon mode
+        then no further actions are possible since chef will have forked into
+        its own process) then a post run function can be executed, which
+        perform finishing activities (such as removing the validation pem
+        file)."""),
```

Comment:
The grammar in the text is a bit broken from the original authors.  Here's another cut that tries to clean it up a bit.

```suggestion
        This module enables chef to be installed (from packages or
        from gems, or from omnibus). Before this occurs chef configurations are
        written to disk (validation.pem, client.pem, firstboot.json,
        client.rb), and needed chef folders/directories are created
        (/etc/chef and /var/log/chef and so-on). Then once installing proceeds
        correctly if configured chef will be started (in daemon mode or in
        non-daemon mode) and then once that has finished (if ran in non-daemon
        mode this will be when chef finishes converging, if ran in daemon mode
        then no further actions are possible since chef will have forked into
        its own process) then a post run function can be executed, which
        perform finishing activities (such as removing the validation pem
        file)."""),
```